### PR TITLE
Adds support for module level test reporting

### LINKED ISSUE #1: ONE_TEST_PER_MODEL does not report module-level errors
Only model- and node-level errors are reported.

### PR DESCRIPTION
**Pull request by wjglerum**
First rough sketch for reporting module level errors

Fixes #1

### CODE REVIEW

**Comment by wjglerum** [comment]
Generated test report from before (notice that it does list the correct number of failures):
```xml
<testsuite name="model checking" tests="0" errors="0" failures="1" time="0" timestamp="2022-10-26T15:51:51"
           hostname="localhost">
    <properties/>
    <system-out></system-out>
    <system-err></system-err>
</testsuite>
```

Generated test report afterwards:
```xml
<testsuite name="model checking" tests="1" errors="0" failures="1" time="0" timestamp="2022-10-26T15:57:54"
           hostname="localhost">
    <properties/>
    <testcase name="module" classname="my.solution.with.module.errors" time="0">
        <failure message="&#xa; Can't find dependency: my.solution.non.existing [my.solution.with.module.errors]"
                 type="model checking"/>
    </testcase>
    <system-out></system-out>
    <system-err></system-err>
</testsuite>
```

File: .gitignore [hunk 1]
```diff
@@ -6,3 +6,7 @@
 # Gradle stuff
 .gradle/
 build/
+
+# MPS stuff
+**/.mps-caches/
+**/.mps/
```
Comment (sergej-koscejev) [inline .gitignore:12]: Don't ignore .mps

File: modelcheck/src/main/kotlin/ModelCheck.kt [hunk 2]
```diff
@@ -262,10 +364,19 @@ fun modelCheckProject(args: ModelCheckArgs, environment: Environment, project: P
         errorCollector.result.map { printResult(it, project, args) }
 
         if (args.xmlFile != null) {
+            val allCheckedModules = itemsToCheck.modules // TODO do we need to filter anything ???
```
Comment (sergej-koscejev) [inline modelcheck/src/main/kotlin/ModelCheck.kt:367]: > do we need to filter anything ???

No, we don't.

File: modelcheck/src/main/kotlin/ModelCheck.kt [hunk 3]
```diff
@@ -12,22 +12,24 @@ import jetbrains.mps.errors.item.IssueKindReportItem
...
 import java.text.SimpleDateFormat
-import java.util.Date
+import java.util.*
 import kotlin.math.min
 import kotlin.test.fail
 
 val logger = Logger.getLogger("de.itemis.mps.gradle.modelcheck")
 
 enum class ReportFormat {
+    @Deprecated("Please use ONE_TEST_PER_MODULE_AND_MODEL instead")
```
Comment (sergej-koscejev) [inline modelcheck/src/main/kotlin/ModelCheck.kt:30]:
```suggestion
    @Deprecated("Please use ONE_TEST_PER_MODULE_AND_MODEL instead, this doesn't report errors on modules.")
```

File: modelcheck/src/main/kotlin/ModelCheck.kt [hunk 4]
```diff
@@ -107,24 +129,61 @@ fun writeJunitXml(models: Iterable<SModel>,
...
+            printError("Model error for [${entry.key.name}]: ${it.message}")
+        }
+    }
+
     val xmlMapper = XmlMapper()
 
+    @Suppress("DEPRECATION")
     when (format) {
         ReportFormat.ONE_TEST_PER_MODEL -> {
             val testcases = oneTestCasePerModel(models, errorsPerModel, project)
```
Comment (sergej-koscejev) [inline modelcheck/src/main/kotlin/ModelCheck.kt:165]: Print a deprecation warning here to STDERR and also to the log.
The warning should tell the user what to do ("Use `--format=module`").

File: modelcheck/src/main/kotlin/ModelCheckArgs.kt [hunk 5]
```diff
@@ -16,10 +16,12 @@ class ModelCheckArgs(parser: ArgParser) : Args(parser) {
...
     val xmlReportFormat by parser.storing("--result-format", help = "reporting format for the JUnit file") {
+        @Suppress("DEPRECATION")
         when (this) {
             "model" -> ReportFormat.ONE_TEST_PER_MODEL
+            "module" -> ReportFormat.ONE_TEST_PER_MODULE_AND_MODEL
             "message" -> ReportFormat.ONE_TEST_PER_FAILED_MESSAGE
             else -> fail("unsupported result format")
         }
-    }.default(ReportFormat.ONE_TEST_PER_MODEL)
+    }.default(ReportFormat.ONE_TEST_PER_MODULE_AND_MODEL)
```
Comment (sergej-koscejev) [inline modelcheck/src/main/kotlin/ModelCheckArgs.kt:26]: Leave the default as it was, for backwards compatibility.

File: modelcheck/src/main/kotlin/ModelCheckArgs.kt [hunk 6]
```diff
@@ -16,10 +16,12 @@ class ModelCheckArgs(parser: ArgParser) : Args(parser) {
     val dontFailOnError by parser.flagging("--error-no-fail", help = "report errors but don't fail the build")
     val xmlFile by parser.storing("--result-file", help = "stores the result as an JUnit xml file").default<String?>(null)
     val xmlReportFormat by parser.storing("--result-format", help = "reporting format for the JUnit file") {
+        @Suppress("DEPRECATION")
         when (this) {
             "model" -> ReportFormat.ONE_TEST_PER_MODEL
+            "module" -> ReportFormat.ONE_TEST_PER_MODULE_AND_MODEL
```
Comment (sergej-koscejev) [inline modelcheck/src/main/kotlin/ModelCheckArgs.kt:22]: Maybe this should be module-and-model or ...?

**Review by sergej-koscejev — CHANGES_REQUESTED** [review]
We also need to bump the minor version (due to --format enhancement) and add some docs about the new things. And the changelog!

File: modelcheck/src/main/kotlin/ModelCheck.kt [hunk 7]
```diff
@@ -219,6 +288,39 @@ private fun oneTestCasePerModel(models: Iterable<SModel>, errorsPerModel: Map<SM
...
+                }
+
+                else -> fail("unexpected issue kind")
+            }
+        }
+
+        val accumulatedFailure = errors.fold(Failure(message = "", type = "model checking"), ::reportItemToContent)
+
+        Testcase(
+            name = "module", // TODO which name to use here?
```
Comment (sergej-koscejev) [inline modelcheck/src/main/kotlin/ModelCheck.kt:316]: `"module " + it.moduleName`

**Review by sergej-koscejev — APPROVED** [review]

**Comment by sergej-koscejev** [comment]
Looks good, merging. Thanks!

**Pull request merged**